### PR TITLE
fix for Chainable RGB LED

### PR DESCRIPTION
**Pull request by tkumata**
Hi,
I use RPi3 and GrovePi+. OS is Raspbian for Robots(wheezy). I connected one Encoder to D2 port and one Chainable RGB LED to D7 port on GrovePi+.
Then I test Chainable RGB LED. So I connect Encoder but do not use Encoder in this test.

I run following simple python script some times as a test.
```
#!/usr/bin/env python
#-*- coding:utf-8 -*-
import time
import grovepi

pin = 7
numleds = 1

grovepi.pinMode(pin,"OUTPUT")
time.sleep(1)

testColorBlack = 0   # 0b000 #000000
testColorBlue = 1    # 0b001 #0000FF
testColorGreen = 2   # 0b010 #00FF00
testColorCyan = 3    # 0b011 #00FFFF
testColorRed = 4     # 0b100 #FF0000
testColorMagenta = 5 # 0b101 #FF00FF
testColorYellow = 6  # 0b110 #FFFF00
testColorWhite = 7   # 0b111 #FFFFFF

thisLedOnly = 0
allLedsExceptThis = 1
thisLedAndInwards = 2
thisLedAndOutwards = 3

try:
    print "Test 1) Initialise"
    grovepi.chainableRgbLed_init(pin, numleds)
    time.sleep(.5)

    print "Test 2a) Test Patterns - black"
    grovepi.chainableRgbLed_test(pin, numleds, testColorBlack)
    time.sleep(1)

    print "Test 2b) Test Patterns - blue"
    grovepi.chainableRgbLed_test(pin, numleds, testColorBlue)
    time.sleep(1)

    print "Test 2c) Test Patterns - green"
    grovepi.chainableRgbLed_test(pin, numleds, testColorGreen)
    time.sleep(1)

    print "Test 2d) Test Patterns - cyan"
    grovepi.chainableRgbLed_test(pin, numleds, testColorCyan)
    time.sleep(1)

    print "Test 2e) Test Patterns - red"
    grovepi.chainableRgbLed_test(pin, numleds, testColorRed)
    time.sleep(1)

    print "Test 2f) Test Patterns - magenta"
    grovepi.chainableRgbLed_test(pin, numleds, testColorMagenta)
    time.sleep(1)

    print "Test 2g) Test Patterns - yellow"
    grovepi.chainableRgbLed_test(pin, numleds, testColorYellow)
    time.sleep(1)

    print "Test 2h) Test Patterns - white"
    grovepi.chainableRgbLed_test(pin, numleds, testColorWhite)
    time.sleep(1)

    quit()
```

Result is following,

### Case 1: firmware v1.2.6 (original)

Test 1) Initialise
Test 2a) Test Patterns - black ... OK
Test 2b) Test Patterns - blue ... OK
Test 2c) Test Patterns - green ... NG, blink black color
Test 2d) Test Patterns - cyan ... NG, blink BLUE color
Test 2e) Test Patterns - red ... NG, blink black color
Test 2f) Test Patterns - magenta ... NG, blink BLUE color
Test 2g) Test Patterns - yellow ... NG, blink black color
Test 2h) Test Patterns - white ... NG, blink BLUE color

Chainable RGB LED can blink only black and blue as stated above.
This test has the similar result on v1.2.3 - v1.2.6.

### Case 2: firmware v1.2.6 (original)

Test 1) Initialise
Test 2a) Test Patterns - black ... OK
Test 2b) Test Patterns - blue ... NG, flash blue color a moment then black soon
Test 2c) Test Patterns - green ... NG, flash green color a moment then black soon
Test 2d) Test Patterns - cyan ... NG, flash cyan color a moment then black soon
Test 2e) Test Patterns - red ... NG, flash red color a moment then black soon
Test 2f) Test Patterns - magenta ... NG, flash magenta color a moment then black soon
Test 2g) Test Patterns - yellow ... NG, flash yellow color a moment then black soon
Test 2h) Test Patterns - white ... NG, flash white color a moment then black soon

I did not find condition for this result yet, but it has result as stated above sometime.

### Case 3: firmware v1.2.6 (origin)

Test 1) Initialise
Test 2a) Test Patterns - black ... OK
Test 2b) Test Patterns - blue ... OK
Test 2c) Test Patterns - green ... OK
Test 2d) Test Patterns - cyan ... OK
Test 2e) Test Patterns - red ... OK
Test 2f) Test Patterns - magenta ... OK
Test 2g) Test Patterns - yellow ... OK
Test 2h) Test Patterns - white ... OK

When I execute cold boot RPi3, it works fine sometimes. But I test continue, it has result of Case 1 or Case 2.

### Case 4: firmware v1.2.6 with my patch
Next, I apply this fix and test.

Test 1) Initialise
Test 2a) Test Patterns - black ... OK
Test 2b) Test Patterns - blue ... OK
Test 2c) Test Patterns - green ... OK
Test 2d) Test Patterns - cyan ... OK
Test 2e) Test Patterns - red ... OK
Test 2f) Test Patterns - magenta ... OK
Test 2g) Test Patterns - yellow ... OK
Test 2h) Test Patterns - white ... OK

Chainable RGB LED works fine even if I test some times.

Best Regards,

### CODE REVIEW

**Comment by DexterInd** [comment]
@karan259 ?

**Comment by tkumata** [comment]
I seem Firmware v1.2.6 has issue. Issue is my comment of pull request. Fix point is following,
https://github.com/DexterInd/GrovePi/pull/247/commits/3ec6fdfe43fe0515f89d80dc229037a84058c973
Best Regards,

File: Firmware/Source/v1.2/grove_pi_v1_2_6/grove_pi_v1_2_6.ino [hunk 1]
```diff
@@ -96,7 +96,7 @@ int j;
 void loop()
 {
   long dur,RangeCm;
-  if(index==4)
+  if(index==4 && flag==0)
```
Comment (nikkoura) [inline Firmware/Source/v1.2/grove_pi_v1_2_6/grove_pi_v1_2_6.ino]: I am not 100% sure, but it seems that at some point in the firmware history, this flag test was removed and replaced with a "run_once" flag. 

This "flag==0" test ensured that all commands would be executed only once.
The "run_once" flag is used to protect only specific commands from continuous repetition.

When patching some time ago the firmware, I was under the impression that this change had been made so as to "spam" commands to the grove sensors (maybe to reduce the impact of i2c communication problems?), except for a select few that would not support such a behavior.

@karan259, could you clarify this point?

**Comment by tkumata** [comment]
Thank you for your reply.
Well, I remove '&& flag==0' from it and I should add following code to block which cmd[0] is 90, 91, 92, 93, 94, 95, right?

if(run_once) {
...
run_once = 0;
}

for example
```
   // Test colors, repeating red green blue
   // color code: 0 black (off), 1 blue, 2 green, 3 cyan, 4 red, 5 magenta, 6 yellow, 7 white
   // [92, pin, num leds, color code]
    else if(cmd[0] == 92)
    {
      if(run_once)
      {
        rgbled[cmd[1]-2].begin(cmd[1], cmd[1]+1, cmd[2]);

        // figure out which color to display, a single bit for each rgb led
        byte rr = ((cmd[3] & 4) >> 2) * 255,
             gg = ((cmd[3] & 2) >> 1) * 255,
             bb = ((cmd[3] & 1)) * 255;

       // set each led to the specified color
        for(int i = 0; i < cmd[2]; i++)
        {
          rgbled[cmd[1]-2].setColorRGB(i, rr, gg, bb);
        }
        run_once = 0;
      }
    }
```

I add above code to block which cmd[0] is 90, 91, 92. I seem Chainable RGB LED works fine.
Best Regards,

**Comment by karan259** [comment]
Hey @tkumata : @nikkoura  is spot on, having **flag==0** makes sure that the commands only run once . I am not sure what **run_once** does. It is same as flag but only applied to smaller set of commands. Were the RGB LED's not working with flag==0??

**Comment by nikkoura** [comment]
Hi @karan259, here is what I gather from the source history:

It seems that the test for `flag == 0` was removed with the release of the firmware in its v1.2.3 (which precisely added support for the RGB LEDs). `flag` is since then correctly set, but never tested. Since that version, commands received by the firmware have been executed in a loop until a new command was received - which can have all sorts of impacts, depending on the way sensors work.

In 1.2.5, the `run_once` variable has been introduced, to achieve something similar to what was once done by the `flag` test, but only for the dust sensor read command. I suppose its internal state might be changed whenever its value is read.

Later on, additional sensors where added that did not worked well with continuous execution of commands, and the `run_once` use has also been used for the temp & humidity sensor, and for the 433MHz transmitter.
I contributed the latter, and stumbled on this issue; not willing to modify the core behavior of the firmware, I used the `run_once` method to limit the changes only to the code I was introducing.

So a question here is: should we re-establish the `flag == 0` test?

**Comment by karan259** [comment]
@nikkoura thanks for the help and @tkumata thanks for the great pull. @DexterInd: tested the pull, pushing a newer script and example too

**Comment by tkumata** [comment]
Hello @karan259,
Thanks for fix.

**Pull request merged**